File: web/packages/doc-runtime/src/vectors.test.ts
import { readdirSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { TimelineDoc, type TimelineOp, type TimelineState } from "./timeline";

// Shared drift vectors: the same files replay through the Go reducer
// (backend/internal/timeline/vectors_test.go) that the export service
// renders from. A change to either reducer that fails here is
// preview/export divergence — fix the port, don't touch the vector.
//
// Replayed through TimelineDoc, NOT bare reduceTimeline: the doc dedups
// op_ids before reducing (as does Go's ParseOps), and the dedup is part of
// the semantics the vectors pin — see dedup-undo-edges.json.

const dir = fileURLToPath(new URL("../../../../spec/timeline-vectors/", import.meta.url));

function normalize(st: TimelineState) {
  return st.tracks.map((t) => ({
    id: t.id,
    kind: t.kind,
    name: t.name,
    clips: t.clips.map((c) => ({
      id: c.id,
      start: c.start,
      duration: c.duration,
      in_point: c.inPoint,
      version_id: c.versionId ?? "",
      shot_id: c.shotId ?? "",
      text: c.text ?? "",
    })),
  }));
}

describe("shared timeline vectors (TS ↔ Go reducer parity)", () => {
  const files = readdirSync(dir).filter((f) => f.endsWith(".json"));
  expect(files.length).toBeGreaterThan(0);
  for (const f of files) {
    const vec = JSON.parse(readFileSync(dir + f, "utf8")) as {
      name: string;
      ops: TimelineOp[];
      expected: { tracks: unknown };
    };
    it(`${f}: ${vec.name}`, () => {
      expect(normalize(new TimelineDoc(vec.ops).state)).toEqual(vec.expected.tracks);
    });
  }
});

File: backend/internal/timeline/vectors_test.go
package timeline

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// Shared drift vectors: the same files replay through the TS reducer
// (web/packages/doc-runtime/src/vectors.test.ts). A change to either
// reducer that fails here is preview/export divergence — fix the port,
// don't touch the vector.

type vecClip struct {
	ID        string  `json:"id"`
	Start     float64 `json:"start"`
	Duration  float64 `json:"duration"`
	InPoint   float64 `json:"in_point"`
	VersionID string  `json:"version_id"`
	ShotID    string  `json:"shot_id"`
	Text      string  `json:"text"`
	Color     *Color  `json:"color"`
}

type vecTrack struct {
	ID    string    `json:"id"`
	Kind  string    `json:"kind"`
	Name  string    `json:"name"`
	Clips []vecClip `json:"clips"`
}

type vector struct {
	Name     string            `json:"name"`
	Ops      []json.RawMessage `json:"ops"`
	Expected struct {
		Tracks []vecTrack `json:"tracks"`
	} `json:"expected"`
}

func TestVectors(t *testing.T) {
	dir := filepath.Join("..", "..", "..", "spec", "timeline-vectors")
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("read vectors dir: %v", err)
	}
	if len(entries) == 0 {
		t.Fatal("no vectors found")
	}
	for _, e := range entries {
		if filepath.Ext(e.Name()) != ".json" {
			continue
		}
		t.Run(e.Name(), func(t *testing.T) {
			raw, err := os.ReadFile(filepath.Join(dir, e.Name()))
			if err != nil {
				t.Fatal(err)
			}
			var v vector
			if err := json.Unmarshal(raw, &v); err != nil {
				t.Fatalf("parse vector: %v", err)
			}
			payloads := make([][]byte, len(v.Ops))
			for i, op := range v.Ops {
				payloads[i] = op
			}
			ops, err := ParseOps(payloads)
			if err != nil {
				t.Fatalf("parse ops: %v", err)
			}
			got := normalize(Reduce(ops))
			if !reflect.DeepEqual(got, v.Expected.Tracks) {
				gj, _ := json.MarshalIndent(got, "", "  ")
				ej, _ := json.MarshalIndent(v.Expected.Tracks, "", "  ")
				t.Errorf("%s: reducer diverged\ngot:  %s\nwant: %s", v.Name, gj, ej)
			}
		})
	}
}

func normalize(st *State) []vecTrack {
	out := make([]vecTrack, 0, len(st.Tracks))
	for _, tr := range st.Tracks {
		vt := vecTrack{ID: tr.ID, Kind: tr.Kind, Name: tr.Name, Clips: []vecClip{}}
		for _, c := range tr.Clips {
			vt.Clips = append(vt.Clips, vecClip{
				ID: c.ID, Start: c.Start, Duration: c.Duration,
				InPoint: c.InPoint, VersionID: c.VersionID, ShotID: c.ShotID,
				Text: c.Text, Color: c.Color,
			})
		}
		out = append(out, vt)
	}
	return out
}
